File: src/main/java/com/kishlaly/ta/analyze/tasks/blocks/two/Long_ScreenTwo_MACD_ThreeBelowZeroAndFigureU.java
package com.kishlaly.ta.analyze.tasks.blocks.two;

import com.kishlaly.ta.model.BlockResultJava;
import com.kishlaly.ta.model.SymbolDataJava;
import com.kishlaly.ta.model.indicators.IndicatorJava;
import com.kishlaly.ta.model.indicators.MACDJava;
import com.kishlaly.ta.utils.LogJava;

import java.util.List;

import static com.kishlaly.ta.model.BlockResultCodeJava.*;

/**
 * histogram should be below zero and start to rise at the last three values
 */
public class Long_ScreenTwo_MACD_ThreeBelowZeroAndFigureU implements ScreenTwoBlockJava {

    @Override
    public BlockResultJava check(SymbolDataJava screen) {
        List<MACDJava> screen_2_MACD = (List<MACDJava>) screen.indicators.get(IndicatorJava.MACD);
        Double macd3 = screen_2_MACD.get(screen_2_MACD.size() - 3).getHistogram();
        Double macd2 = screen_2_MACD.get(screen_2_MACD.size() - 2).getHistogram();
        Double macd1 = screen_2_MACD.get(screen_2_MACD.size() - 1).getHistogram();

        boolean histogramBelowZero = macd3 < 0 && macd2 < 0 && macd1 < 0;
        if (!histogramBelowZero) {
            LogJava.recordCode(HISTOGRAM_NOT_BELOW_ZERO_SCREEN_2, screen);
            LogJava.addDebugLine("The histogram on the second screen is at least zero");
            return new BlockResultJava(screen.getLastQuote(), HISTOGRAM_NOT_BELOW_ZERO_SCREEN_2);
        }

        boolean figureU = macd2 < macd3 && macd2 < macd1;
        if (!figureU) {
            LogJava.recordCode(HISTOGRAM_NOT_ASCENDING_SCREEN_2, screen);
            LogJava.addDebugLine("The histogram on the second screen does not form a negative U");
            return new BlockResultJava(screen.getLastQuote(), HISTOGRAM_NOT_ASCENDING_SCREEN_2);
        }
        return new BlockResultJava(screen.getLastQuote(), OK);
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/tasks/blocks/two/Long_ScreenTwo_MACD_ThreeBelowZeroAndFigureU.kt
package com.kishlaly.ta.analyze.tasks.blocks.two

import com.kishlaly.ta.model.BlockResult
import com.kishlaly.ta.model.BlockResultCode
import com.kishlaly.ta.model.SymbolData
import com.kishlaly.ta.model.indicators.MACD
import com.kishlaly.ta.utils.Log
import com.kishlaly.ta.utils.SymbolDataUtils

/**
 * histogram should be below zero and start to rise at the last three values
 */
class Long_ScreenTwo_MACD_ThreeBelowZeroAndFigureU : ScreenTwoBlock {
    override fun check(screen: SymbolData): BlockResult {
        val macd = SymbolDataUtils(screen, MACD::class.java)
        val histogramBelowZero = macd.last(3).histogram < 0
                && macd.last(2).histogram < 0
                && macd.last().histogram < 0
        if (!histogramBelowZero) {
            Log.recordCode(BlockResultCode.HISTOGRAM_NOT_BELOW_ZERO_SCREEN_2, screen)
            Log.addDebugLine("The histogram on the second screen is at least zero")
            return BlockResult(screen.lastQuote, BlockResultCode.HISTOGRAM_NOT_BELOW_ZERO_SCREEN_2)
        }

        val figureU = macd.last(2).histogram < macd.last(3).histogram
                && macd.last(2).histogram < macd.last().histogram
        if (!figureU) {
            Log.recordCode(BlockResultCode.HISTOGRAM_NOT_ASCENDING_SCREEN_2, screen)
            Log.addDebugLine("The histogram on the second screen does not form a negative U")
            return BlockResult(screen.lastQuote, BlockResultCode.HISTOGRAM_NOT_ASCENDING_SCREEN_2)
        }
        return BlockResult(screen.lastQuote, BlockResultCode.OK)

    }
}